graph/telemetry/istio/appender/health.go

```diff
@@ -105,18 +106,96 @@ func (a *HealthAppender) attachHealth(trafficMap graph.TrafficMap, globalInfo *g
 	// Execute health fetches and attach retrieved health data to nodes
 	bs := globalInfo.Business
 	ctx := globalInfo.Context
+
+	// Gather all the health data we'll need ahead of time key'd by namespace.
+	appHealth := make(map[string]models.NamespaceAppHealth)
+	serviceHealth := make(map[string]models.NamespaceServiceHealth)
+	workloadHealth := make(map[string]models.NamespaceWorkloadHealth)
+	type result struct {
+		namespace      string
+		appHealth      *models.NamespaceAppHealth
+		serviceHealth  *models.NamespaceServiceHealth
+		workloadHealth *models.NamespaceWorkloadHealth
+		err            error
+	}
+	healthCh := make(chan result)
+	var errors []error
+
+	// Start accumulating results before fetching health info
+	doneAccumulating := make(chan struct{})
+	go func() {
+		// Need to loop over healthCh until closed, otherwise any go routines
+		// sending to this chan will remain blocked forever. Ideally as soon
+		// as we get an error we'd propagate cancellation to all the running
+		// goroutines.
+		for res := range healthCh {
+			if res.err != nil {
+				errors = append(errors, res.err)
+			}
+			if res.appHealth != nil {
+				appHealth[res.namespace] = *res.appHealth
+			} else if res.serviceHealth != nil {
+				serviceHealth[res.namespace] = *res.serviceHealth
+			} else if res.workloadHealth != nil {
+				workloadHealth[res.namespace] = *res.workloadHealth
+			}
+		}
+		doneAccumulating <- struct{}{}
+	}()
```

Comment:
I'd lean towards not doing parallelization of this block.
Since this is doing simple assignments, I think time reduction will be negligible.

I think that synchronization could be achieved purely by using the `healthCh`, removing the need of the `wg` WaitGroup and the `doneAccumulating` channel, making code slightly clearer. Basically, rather than using `wg.Add(i)` use a simple counter to know how many thread you've created. Then, move this block of code down after the `wg.Wait()`. You will no longer need to enclose this block inside the `go func()`. Then, the `for res := range healthCh {` would become a simpler `for i := 1; i <= goroutines_counter; i++ {`

Reply:
Yeah that'd work. Basically doing?:
```
	for i := 0; i <= numberOfRequests; i++ {
		res := <-healthCh	
          ....
	}
```

Reply:
+1, parallel operations should involve typically I/O, this goroutine can go in the main flow.

Reply:
> Yeah that'd work. Basically doing?:
> 
> ```
> 	for i := 0; i <= numberOfRequests; i++ {
> 		res := <-healthCh	
>           ....
> 	}
> ```

You've got the idea :smile: